modules/ruxhal/src/virtio/virtio_console.rs

```diff
@@ -0,0 +1,218 @@
+    buffer: [u8; 20000],
+    #[cfg(feature = "irq")]
+    irq_buffer: SpinNoIrq<RxRingBuffer>,
+    pointer: usize,
+    addr: usize,
+}
+
+/// The UART driver instance
+static mut UART: UartDrv = UartDrv {
+    inner: None,
+    buffer: [0; 20000],
+    #[cfg(feature = "irq")]
+    irq_buffer: SpinNoIrq::new(RxRingBuffer::new()),
+    pointer: 0,
+    addr: 0,
+};
+
+/// Writes a byte to the console.
+pub fn putchar(c: u8) {
+    unsafe {
+        if let Some(ref mut uart_inner) = UART.inner {
+            if UART.pointer > 0 {
+                for i in 0..UART.pointer {
+                    let mut uart = uart_inner.lock();
+                    match UART.buffer[i] {
+                        b'\n' => {
+                            uart.putchar(b'\r');
+                            uart.putchar(b'\n');
+                        }
+                        c => uart.putchar(c),
+                    }
+                }
+                UART.pointer = 0;
+                warn!("######################### The above content is printed from buffer! #########################");
+            }
+            let mut uart = uart_inner.lock();
+            match c {
+                b'\n' => {
+                    uart.putchar(b'\r');
+                    uart.putchar(b'\n');
+                }
+                c => uart.putchar(c),
+            }
+        } else {
+            UART.buffer[UART.pointer] = c;
+            UART.pointer += 1;
+        }
+    }
+}
+
+/// Reads a byte from the console.
+pub fn getchar() -> Option<u8> {
+    unsafe {
+        #[cfg(feature = "irq")]
+        return UART.irq_buffer.lock().pop();
+        #[cfg(not(feature = "irq"))]
+        if let Some(ref mut uart_inner) = UART.inner {
+            let mut uart = uart_inner.lock();
+            return uart.getchar();
+        } else {
+            None
+        }
+    }
+}
+
+/// probe virtio console directly
+pub fn directional_probing() {
+    info!("Initiating VirtIO Console ...");
+    for reg in ruxconfig::VIRTIO_MMIO_REGIONS {
+        {
+            if let Some(dev) = probe_mmio(reg.0, reg.1) {
+                unsafe {
+                    UART.inner = Some(SpinNoIrq::new(dev));
+                    UART.addr = reg.0;
+                }
+            }
+        }
+    }
+    info!("Output now redirected to VirtIO Console!");
+}
+
+/// enable virtio console interrupt
+pub fn enable_interrupt() {
+    #[cfg(all(feature = "irq", target_arch = "aarch64"))]
+    {
+        info!("Initiating VirtIO Console interrupt ...");
+        info!("IRQ ID: {}", crate::platform::irq::UART_IRQ_NUM);
+        crate::irq::register_handler(crate::platform::irq::UART_IRQ_NUM, irq_handler);
+        crate::irq::set_enable(crate::platform::irq::UART_IRQ_NUM, true);
+        ack_interrupt();
+        info!("Interrupt enabled!");
+    }
+}
+
+/// virtio console interrupt handler
+pub fn irq_handler() {
+    #[cfg(feature = "irq")]
+    unsafe {
```

Comment:
nesting too much uncessary brackets, another question is that this `unsafe` is avoidable.